lib/metrics.py

```diff
@@ -348,3 +358,33 @@ def _get_db_conn(self):
                 psycopg2.extensions.ISOLATION_LEVEL_AUTOCOMMIT
             )
         return self.db
+
+
+class PaidAppsMetricsEmitterThread(BaseMetricsEmitterThread):
+    def _selected_stats_to_emit(self):
+        selected_stats = []
+        if buildpackutil.i_am_primary_instance():
+            selected_stats = [
+                self._inject_database_stats,
+                self._inject_storage_stats,
+                self._inject_health,
+            ]
+        selected_stats.append(self._inject_m2ee_stats)
+        return selected_stats
+
+
+class FreeAppsMetricsEmitterThread(BaseMetricsEmitterThread):
+    def _extract_session_metrics(self, stats):
+
+        try:
+            session_metrics = stats["mendix_runtime"]["sessions"]
```

Comment:
Instead of stripping out metrics after the fact, wouldn't it be simpler to only emit the stats that actually need emitting by overriding the run method?

Reply:
Do you mean making changes to m2ee module?  This method is executed before emmiting metrics not after.
All runtime metrics are fetched via `munin.get_stats_from_runtime` in` _inject_m2ee_stats()` then only session metrics are extracted with this method. Then metrics get emmited.

Reply:
No, I mean something a bit different:

1) We call `BaseMetricsEmitterThread._inject_m2ee_stats`, which itself calls out to `munin`, `m2ee.client.get_critical_log_messages`, and a few other things. But we only want a subset of those things. I think that it would be better to simply add a `_get_user_sessions_stats` method to get the user sessions stats, and not, for example, the critical log messages count. 

    If you wanted to call this new method from `_inject_m2ee_stats`_, that would be fine.

2) In `BaseMetricsEmitterThread.run,  we loop over all selected stats methods to emit, and then call those methods. Depending on if there were exceptions thrown during these methods, we then add health information; either good or bad. However, health information isn't relevant for free apps. I think that the try/except calls should be different for free apps and for paid apps. For paid apps, they should remain the same. For free apps, we should simply do nothing in the case of an exception.

Reply:
Well. I wanted to not touch the original code as much as possible. But since we are on it.

1. Maybe you're right here, also after taking into account what Chris mentioned about nasty side effects if `extract_session_metrics` is the last called method. I will create `_get_user_sessions_stats` method instead using _inject_m2ee_stats_ followed by `_extract_session_metrics`. 

2. In my opinion, all the _inject methods should catch their own exceptions and handle them accordingly so that there should be no need for "try exception" clause in the run method at all. `run` should just be a dummy wrapper, no logic, no test required. imho.
The fact that we degrade the health status of the application just because there was ANY exception during gathering metrics is weird. Also, any successfully gathered metrics are lost due to the fact that we are simply replacing the stats dictionary, instead of replacing the value(s) under "health" key

All of those changes mean that I should introduce more tests to the project.

Reply:
I strongly agree with not touching the original code as much as possible.

I think that for the sake of simplicity, you can create your own simpler `.run` method on the `FreeAppsMetricEmitter` class, and move the `.run` method on `BaseAppsMetricEmitter` to `PaidAppsMetricEmitter`, but leaving the code as it was before your changes.

This means we can keep free apps elegant, but without changing anything for existing apps.

Re: point 2, I agree that this is weird (indeed, this is the source of some annoying problems with alerting). But, like your first point, I would again change this as little as is possible.

Reply:
>I will create _get_user_sessions_stats method instead using inject_m2ee_stats followed by _extract_session_metrics.

I would do this differently:

```python
    def _inject_free_apps_session_stats(self, stats):
        m2ee_stats, _ = munin.get_stats_from_runtime(
            self.m2ee.client, self.m2ee.config
        )
        stats["mendix_runtime"] = {}
        stats["mendix_runtime"]["sessions"] = m2ee_stats["sessions"]
        return stats

    # This is unchanged from the current situation
    def _inject_m2ee_stats(self, stats):
        m2ee_stats, java_version = munin.get_stats_from_runtime(
            self.m2ee.client, self.m2ee.config
        )
        if "sessions" in m2ee_stats:
            m2ee_stats["sessions"]["user_sessions"] = {}
        m2ee_stats = munin.augment_and_fix_stats(
            m2ee_stats, self.m2ee.runner.get_pid(), java_version
        )

        critical_logs_count = len(self.m2ee.client.get_critical_log_messages())
        m2ee_stats["critical_logs_count"] = critical_logs_count
        stats["mendix_runtime"] = m2ee_stats
        return stats
```

This would cut down on the number of m2ee calls for free apps, which would IMO be preferable.

Reply:
How is that different from
>I will create _get_user_sessions_stats method instead using inject_m2ee_stats followed by _extract_session_metrics. 
??
Also why don't you just commit this to this PR? This discussion is going to some weird place.